lib/BuildSystem/LaneBasedExecutionQueue.cpp

```diff
@@ -329,6 +405,29 @@ class LaneBasedExecutionQueue : public BuildExecutionQueue {
                                          status);
     return (status == 0);
   }
+
+  // Reads a fd and executes callback on each read 
+  // until either EOF is encountered or read() returns 
+  // with an error.
+  // Returns true if EOF is found.
+  // Returns false if read() returns negative value.
+  bool readFd(int& fd, 
+              std::function<void(StringRef)> callback) {
+    while (true) {
+      char buf[4096];
+      ssize_t numBytes = read(fd, buf, sizeof(buf));
+      if (numBytes < 0) {
+        return false;
+      }
+
+      if (numBytes == 0)
+        break;
+      if (callback) {
+        callback(std::string(buf, numBytes));
```

Comment:
Use `StringRef()` here, not `std::string` (which will create an unnecessary allocation).